### ISSUE TITLE
Pagination bug in GET /projects

### ISSUE THREAD

**Issue by JWCook** [inaturalist/iNaturalistAPI#227]
There appears to be an issue with paginating responses from `GET /projects`.

# TL;DR
When using a `per_page` value larger than 30, and a `page` value larger than 1, the **results are only offset by 30**, not by `per_page`. In other words, a total of `page_size - 30` results will be duplicated between page `n` and page `n+1`.

# Example
Here is an example of this behavior using python. Start by fetching 2 pages of results:
```python
import requests
BASE_URL = 'https://api.inaturalist.org/v1/projects'

response_page_1 = requests.get(BASE_URL, params={'q': 'invasive', 'per_page': 200, 'page': 1})
response_page_2 = requests.get(BASE_URL, params={'q': 'invasive', 'per_page': 200, 'page': 2})
ids_page_1 = [p['id'] for p in response_page_1.json()['results']]
ids_page_2 = [p['id'] for p in response_page_2.json()['results']]

print(
    f"Total results: {response_page_1.json()['total_results']}; "
    f"Page 1 results: {len(ids_page_1)}; page 2 results: {len(ids_page_2)}"
)
# Total results: 339; Page 1 results: 200; page 2 results: 200
```

For a page size of 200 and 339 total results, expected behavior would be 139 results on page 2. Counting the unique project IDs shows the issue:
```python
unique_ids = set(ids_page_1 + ids_page_2)
print(len(unique_ids))
# 230
```

So page 2 returned 170 of the same results as page 1, plus 30 new results. There are no duplicate results within a given page. The behavior is the same for subsequent pages:
```python

response_page_3 = requests.get(BASE_URL, params={'q': 'invasive', 'per_page': 200, 'page': 3})
ids_page_3 = [p['id'] for p in response_page_3.json()['results']]
len(set(ids_page_1 + ids_page_2 + ids_page_3))
# 260
```

This appears consistent with SQL results from `LIMIT $per_page OFFSET ($page-1) * 30` rather than `LIMIT $per_page OFFSET ($page-1) * $per_page `. The [Projects search page](https://www.inaturalist.org/projects) in the web app has a page size of 30, so I'm guessing this offset is hardcoded somewhere.

**Comment by JWCook** [comment]
P.S., This appears to be the same issue as #218  but for a different endpoint.

**Comment by kueda** [comment]
@JWCook I think this fix was deployed. Can you check to see if it's working for you now?

**Comment by JWCook** [comment]
@kueda Thanks for the response! Results are closer to what's exepcted, but something is still off. My example above with 2 pages works correctly, but other page sizes return varying numbers of duplicate results:

```python
import requests
BASE_URL = 'https://api.inaturalist.org/v1/projects'

def get_paginated_results(per_page, n_pages):
    ids = []
    for i in range(n_pages):
        response = requests.get(BASE_URL, params={'q': 'invasive', 'per_page': per_page, 'page': i
 + 1})
        ids.extend([p['id'] for p in response.json()['results']])
    unique_ids = set(ids)

    print(f"Total results: {response.json()['total_results']}")
    print(f"Total IDs received: {len(ids)}")
    print(f"Unique IDs received: {len(unique_ids)}")
```

Results from 8 different page sizes:
```python
>>> get_paginated_results(200, 2)
Total results: 342
Total IDs received: 342
Unique IDs received: 342

>>> get_paginated_results(150, 3)
Total results: 342
Total IDs received: 342
Unique IDs received: 335

>>> get_paginated_results(100, 4)
Total results: 342
Total IDs received: 342
Unique IDs received: 333

>>> get_paginated_results(75, 5)
Total results: 342
Total IDs received: 342
Unique IDs received: 337

>>> get_paginated_results(50, 7)
Total results: 342
Total IDs received: 342
Unique IDs received: 323

>>> get_paginated_results(30, 12)
Total results: 342
Total IDs received: 342
Unique IDs received: 320

>>> get_paginated_results(20, 18)
Total results: 342
Total IDs received: 342
Unique IDs received: 313

>>> get_paginated_results(10, 35)
Total results: 342
Total IDs received: 342
Unique IDs received: 284
```

Condensed (per_page, # of pages, total unique IDs, # of duplicate results):
```python
200x2: 342  |  0
150x3: 335  |  7
100x4: 333  |  9
75x5:  337  |  5
50x7:  323  |  19
30x12: 320  |  22
20x18: 313  |  29
10x35: 284  |  58
```

I can't immediately identify the pattern there, but I'll take another look later this week.

**Comment by JWCook** [comment]
Okay, I just ran those same tests again and got different results:
```python
200x2: 342  |  0
150x3: 334  |  8
100x4: 340  |  2
75x5:  338  |  4
50x7:  315  |  22
30x12: 324  |  18
20x18: 301  |  41
10x35: 286  |  52
```

Ran it a couple more times and got different results each time. I don't have much experience with Node.js, but looking over `ProjectsController` I don't see anything there that could be responsible for nondeterministic behavior.

Looks like that's hitting an Elasticsearch database. Is it possible that there's a problem with the search index?

**Comment by kueda** [comment]
@pleary, I ran the Python code above on staging and could replicate the output, though I can in production. I suspect @JWCook is right and we have some duplicate documents in our projects index, or some other weirdness happening. We could reindex all projects, but that won't clear out leftover documents in the index that were never properly deleted, will it?

**Comment by pleary** [comment]
I don't think we have duplicate documents in the index. Rather I think this is just an unfortunate known behavior with Elasticsearch with multiple shards when using default scoring (see https://www.elastic.co/guide/en/elasticsearch/reference/master/consistent-scoring.html). One way right now to get better consistency across pages is to use one of the `order_by` search parameters. I tried adding that to the python script and in a few tests saw no duplicates. 

If we rebuilt the index, the results would likely be consistent across all shards at first but would eventually drift as each shard would do different things with deleted documents.

So I'd recommend using an `order_by` param, maybe `order_by=created` and use the highest `per_page` value for your use-case. Keep in mind that the data are dynamic, so even if Elasticsearch were 100% consistent, projects get added, deleted, and modified all the time which throw off which results show on which page. Thus we could never guarantee any kind of stability across multiple requests.

**Comment by JWCook** [comment]
Thanks for the clarification! I can confirm that using `order_by` appears to give consistent results without duplicates.

Would this potentially affect any endpoint with a text search index (`'q'` parameter)? If so, it may be worth adding a note to that effect to the API docs and/or [API Recommended Practices](https://www.inaturalist.org/pages/api+recommended+practices) page.

Otherwise, I'd consider this fixed and you can close the issue if you like.

**Comment by JWCook** [comment]
Could your fix for this issue also be applied to `/projects/autocomplete` (issue #218)? In `projects_controller.autocomplete()` on [line 176](https://github.com/inaturalist/iNaturalistAPI/blob/main/lib/controllers/v1/projects_controller.js#L176), possibly?
